pkg/loadbalancer/reflectors/conversions.go

```diff
@@ -0,0 +1,386 @@
+						fes = append(fes, fe)
+					}
+				}
+
+			}
+		}
+
+		// ExternalIP
+		for _, ip := range svc.Spec.ExternalIPs {
+			addr, err := cmtypes.ParseAddrCluster(ip)
+			if err != nil {
+				continue
+			}
+			if (!cfg.EnableIPv6 && addr.Is6()) || (!cfg.EnableIPv4 && addr.Is4()) {
+				continue
+			}
+
+			for _, port := range svc.Spec.Ports {
+				fe := loadbalancer.FrontendParams{
+					Type:        loadbalancer.SVCTypeExternalIPs,
+					PortName:    loadbalancer.FEPortName(port.Name),
+					ServiceName: name,
+					ServicePort: uint16(port.Port),
+				}
+
+				p := loadbalancer.NewL4Addr(loadbalancer.L4Type(port.Protocol), uint16(port.Port))
+				if p == nil {
+					continue
+				}
+
+				fe.Address.AddrCluster = addr
+				fe.Address.Scope = loadbalancer.ScopeExternal
+				fe.Address.L4Addr = *p
+				fes = append(fes, fe)
+			}
+		}
+	}
+
+	return
+}
+
+func getIPFamilies(svc *slim_corev1.Service) []slim_corev1.IPFamily {
+	if len(svc.Spec.IPFamilies) == 0 {
+		// No IP families specified, try to deduce them from the cluster IPs
+		if len(svc.Spec.ClusterIP) == 0 || svc.Spec.ClusterIP == slim_corev1.ClusterIPNone {
+			return nil
+		}
+
+		ipv4, ipv6 := false, false
+		if len(svc.Spec.ClusterIPs) > 0 {
+			for _, cip := range svc.Spec.ClusterIPs {
+				if ip.IsIPv6(net.ParseIP(cip)) {
+					ipv6 = true
+				} else {
+					ipv4 = true
+				}
+			}
+		} else {
+			ipv6 = ip.IsIPv6(net.ParseIP(svc.Spec.ClusterIP))
+			ipv4 = !ipv6
+		}
+		families := make([]slim_corev1.IPFamily, 0, 2)
+		if ipv4 {
+			families = append(families, slim_corev1.IPv4Protocol)
+		}
+		if ipv6 {
+			families = append(families, slim_corev1.IPv4Protocol)
+		}
+		return families
+	}
+	return svc.Spec.IPFamilies
+}
+
+func convertEndpoints(cfg loadbalancer.ExternalConfig, ep *k8s.Endpoints) (name loadbalancer.ServiceName, out []loadbalancer.BackendParams) {
+	name = loadbalancer.ServiceName{
+		Name:      ep.ServiceID.Name,
+		Namespace: ep.ServiceID.Namespace,
+	}
+
+	// k8s.Endpoints may have the same backend address multiple times
+	// with a different port name. Collapse them down into single
+	// entry.
+	type entry struct {
+		portNames []string
+		backend   *k8s.Backend
+	}
+	entries := map[loadbalancer.L3n4Addr]entry{}
+
+	for addrCluster, be := range ep.Backends {
+		if (!cfg.EnableIPv6 && addrCluster.Is6()) || (!cfg.EnableIPv4 && addrCluster.Is4()) {
+			continue
+		}
+		for portName, l4Addr := range be.Ports {
+			l3n4Addr := loadbalancer.L3n4Addr{
+				AddrCluster: addrCluster,
+				L4Addr:      *l4Addr,
+			}
+			if isIngressDummyEndpoint(l3n4Addr) {
+				continue
+			}
+			portNames := entries[l3n4Addr].portNames
+			if portName != "" {
+				portNames = append(portNames, portName)
+			}
+			entries[l3n4Addr] = entry{
+				portNames: portNames,
+				backend:   be,
+			}
+		}
+	}
+	for l3n4Addr, entry := range entries {
+
+		state := loadbalancer.BackendStateActive
+		if entry.backend.Terminating {
+			state = loadbalancer.BackendStateTerminating
```

Comment:
is this on purpose or do we not support other states than active/terminating?

Reply:
From k8s types these are the only 2 states we can reach. The `Quarantined` state can be reached with a health checker that marks a backend as unhealthy (currently there are none in OSS cilium), and the `Maintenance` is currently completely unused (this could've been set via the `cilium-dbg service update` that is now being removed).

Reply:
yeah `Maintenance` is probably usefull only for the standalone mode (to set backend for maintenance), could we support this via annotation potentially?

Reply:
yeah, just wanted to mention weights

Reply:
Yep we could consider adding an annotation for both maintenance and for weights. Were weights needed for your use-case?

Reply:
yeah we have some backends that are not as performant as others